docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,247 @@
+    trace = k.GetTrace(ctx, traceHash)
+    // prefix the identifiers to create the new trace
+    trace = portID + "/" + channelID +"/" + trace + "/"
+  }
+  
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String()  "/" + baseDenom
+  return denom
+}
+```
+
+The denomination also needs to be updated when token is received on the source chain:
+
+```golang
+// UnprefixDenom removes the first portID/channelID pair from a given denomination trace info and returns the
+// denomination with the updated trace hash and the new trace info.
+// An error is returned if the trace cannot be found on the store from the denom's trace hash.
+func (k Keeper) UnprefixDenom(ctx Context, denom string) (denom, trace string, err error) {
+  denomSplit := strings.Split(denom, "/")
+  if denomSplit[0] == denom {
+    return denom, fmt.Errorf("denomination %s doesn't contain a prefix", denom)
+  }
+
+  // concatenate if base denom already contains a separator
+  baseDenom = strings.Join(denomSplit[2], "/")
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> denom identifiers prefix
+  trace = k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  traceSplit := strings.Split(trace, "/")
+  if len(traceSplit) == 3 {
+    // the trace has only one portID/channelID pair
+    return baseDenom
+  }
+
+  // remove a single identifiers pair to create the new trace
+  trace = strings.Join(trace[2:], "/")
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String() + "/" + baseDenom
+  return denom
+}
+```
+
+```golang
+// GetTraceFromDenom returns the token source tracing info from the trace hash of the given denomination
+func (k Keeper) GetTraceFromDenom(ctx Context, denom string) (string, error) {
+  denomSplit := strings.Split(denom, "/")
+  if denomSplit[0] == denom {
+    return "", nil
+  }
+
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> denom identifiers prefix
+  trace := k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  return trace
+}
+
+```
+
+Additionally, the `SendTransfer`'s `createOutgoingPacket` call and the `OnRecvPacket` need to be updated to be retreive the trace info (using `GetTraceFromDenom`) prior to checking the correctness of the prefix.
+
+### Coin Changes
+
+The coin denomination validation will need to be updated to reflect these changes. In particular, the denomination validation
+function will now accept slash separators (`"/"`) and will bump the maximum character length to 64.
```

Comment:
I think it already has a maximum character length of 64?

Reply:
correct, on master, but it hasn't been included on any release 👍